### DISCUSSION TITLE
Berry: what is best way to save variable/data persitant

### DISCUSSION THREAD

**Discussion by ChrisVonFrosta** [arendst/Tasmota#18372] [General]
I would like to persistently store a counter value in BerryScript across reboots or power-on cycles, either by periodically saving it (every minute) or by using a brown-out-detection routine. However, due to frequent write operations, using flash memory may not be advisable. What options do I have in BerryScript? Can I use a filesystem? Is there a brown-out detection mechanism available?
i would prefer using buildin flash storage if available. But i do care about the 1440 write cycles per day (24*60).

**Comment by barbudor** [comment]
Berry has a feature for that : https://tasmota.github.io/docs/Berry/#persist-module
persistant data are stored on the filesystem in file `_persist.json`

When writing into a file, data are not overwriting their previous location but a new sector is used at each write and thus it spread the wearing across all free sectors. So thta shouldn't be much of a problem.

Alternatively a I2C EEPROM would have millions of write cycle per byte
Or even the NVRAM section of a RTC chip such as the DS3231 where data are saved by the backup battery. NVRAM have no limits to writing. Using a supercap instead of a battery and your are good for years of service.

Of course wiht Tasmota intended to be connected to MQTT, you still have the ability to use MQTT to save your counter. If you publish to a retained topic and self-subscribe to that topic you would get back what you have last written on a restart.

**Reply by s-hadinger** [reply]
Does littlefs actually use new blocks for wear leveling? Tasmota doesn't implement anything like such for file system

**Reply by barbudor** [reply]
I didn't checked specifically for littlefs but most filesystem behave like that : write in a new location before marking the previous one free
